### PR TITLE
Issue #13934: Version updated from 3.2 to 3.4

### PR DESCRIPTION
**Pull request by Rohanraj123**
related #13934

### CODE REVIEW

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java [hunk 1]
```diff
@@ -164,7 +164,7 @@ public void setFormat(Pattern pattern) {
      * in non-void methods/lambdas.
      *
      * @param max maximum allowed number of return statements.
-     * @since 3.2
+     * @since 3.4
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java:167]: See previous comment. You can see max setter in it.

**Review by rnveach — CHANGES_REQUESTED** [review]

**Comment by rnveach** [comment]
From issue: https://github.com/checkstyle/checkstyle/pull/13932/files#r1368105661

Property was `format` originally. It was changed from 3.4 to 3.2 originally.

https://github.com/checkstyle/checkstyle/blob/release3_4/src/checkstyle/com/puppycrawl/tools/checkstyle/checks/AbstractFormatCheck.java

3.2 check above didn't specify `format` or extend `AbstractFormatCheck`.

It did extend `AbstractFormatCheck` in 3.4 .
https://github.com/checkstyle/checkstyle/blob/release3_4/src/checkstyle/com/puppycrawl/tools/checkstyle/checks/coding/ReturnCountCheck.java

But it did in 3.4 .
https://github.com/checkstyle/checkstyle/blob/release3_4/src/checkstyle/com/puppycrawl/tools/checkstyle/checks/AbstractFormatCheck.java

**Comment by Rohanraj123** [comment]
@romani Sir can you please review the changes and please provide me some enhancements if required.

**Comment by rnveach** [comment]
@Rohanraj123 You have not resolved all my review items yet.

**Comment by romani** [comment]
@Rohanraj123 , please redirect all questions to @rnveach , he is first reviewer for this PR.

**Comment by Rohanraj123** [comment]
> @Rohanraj123 You have not resolved all my review items yet.

@rnveach Sir, can you please one more time define the doables so that i can keep them in mind while making the changes. Would be helpful for my learning as well.

**Comment by rnveach** [comment]
@Rohanraj123 Where is the confusion with the information you were given at https://github.com/checkstyle/checkstyle/issues/13934#issuecomment-1777352322 and my review items at https://github.com/checkstyle/checkstyle/pull/13945#discussion_r1370267758 , https://github.com/checkstyle/checkstyle/pull/13945#discussion_r1370270679 , and https://github.com/checkstyle/checkstyle/pull/13945#issuecomment-1777336667 .

**Comment by rnveach** [comment]
@Rohanraj123 
The line at https://github.com/checkstyle/checkstyle/blob/0322ba1be8489d1df373013efc9078de4d209eb4/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java#L702 needs to be removed.

With it gone, it will re-generate the XML files with the correct versions.

**Comment by Rohanraj123** [comment]
> @Rohanraj123 The line at
> 
> https://github.com/checkstyle/checkstyle/blob/0322ba1be8489d1df373013efc9078de4d209eb4/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java#L702
> 
> needs to be removed.
> With it gone, it will re-generate the XML files with the correct versions. I recommend to start with a clean master first.

Sure sir Let me do the changes

**Comment by Rohanraj123** [comment]
@rnveach Sir, removing the line of code you were reffering giving error : ```
[ERROR] com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest -- Time elapsed: 2.495 s <<< ERROR!
java.lang.NullPointerException: Cannot invoke "com.puppycrawl.tools.checkstyle.api.DetailNode.getChildren()" because "node" is null
```

**Comment by rnveach** [comment]
Update the PR. I would need to see what is being done and the full stack trace.

**Comment by Rohanraj123** [comment]
@rnveach Sir, I have added the catch check in XdocGenerator file. And squash them in one commit. When i tried to run mvn clean verify , It says that Exception arrives in src\xdocs\checks\whitespace\filetabcharacter.xml.template .
Sir it might take some time for me to work on this issue because my university exams are going on. So Please try to provide me some time.

**Review by romani — APPROVED** [review]
Ok to merge

**Comment by rnveach** [comment]
@Rohanraj123 No worries. You have some time. I am looking into the issue a bit more to understand things.

@romani CI is failing, so is too soon to say it is ready for merging.

**Comment by rnveach** [comment]
I was digging through the error and I think I understand the logic of why the code we removed is there. `filetabcharacter` is an `FileSetCheck`. The error for the since version is because it is looking for the javadoc of `fileExtensions` and not finding it since it is in the parent class. This is a property for all `FileSetCheck`, so I assume the reasoning for the removed code was that this class has been here long enough that all modules that use it, should have had it from the start. However, this logic starts to fall apart if we add a new property in the future to any abstract checks that are already in existence.

We can actually see this for 2 of the overrides we did for `SINCE_VERSION_FOR_INHERITED_PROPERTY`:
https://github.com/checkstyle/checkstyle/blob/1282fe6a9e07a7b2ec76e0a86144e55899294c35/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java#L165-L166

And I think that's where the solution lies, we need to add these properties to this global override field. So I would expect us to add a few `<ModuleName>.fileExtensions` to the list with each module's specific version that it defines. Most likely all may be the same as Module's own version number.

Ideally, we could remove the need for this list if we did version comparing between the version the module was introduced versus the version the inherited property was introduced and taking the newest of the 2. We are tracking inherited properties directly through `SUPER_CLASS_PROPERTIES_JAVADOCS`. I would expect this to be done in another PR. @romani If you agree we should create the issue for this.

**Comment by romani** [comment]
@Rohanraj123, if you think that this PR becomes a bit complicated to finish.
It is completely fine to pause it or abandon it and switch to easier issues to get experience a bit more before handling complicated.
We want you to be productive and not be burned by heavy problems.

**Comment by Rohanraj123** [comment]
> I was digging through the error and I think I understand the logic of why the code we removed is there. `filetabcharacter` is an `FileSetCheck`. The error for the since version is because it is looking for the javadoc of `fileExtensions` and not finding it since it is in the parent class and has no version number defined there. This is a property for all `FileSetCheck`, so I assume the reasoning for the removed code was that this class has been here long enough that all modules that use it, should have had it from the start. However, this logic starts to fall apart if we add a new property in the future to any abstract checks that are already in existence.
> 
> We can actually see this for 2 of the overrides we did for `SINCE_VERSION_FOR_INHERITED_PROPERTY` for 2 same inherited properties:
> 
> https://github.com/checkstyle/checkstyle/blob/1282fe6a9e07a7b2ec76e0a86144e55899294c35/src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java#L165-L166
> 
> And I think that's where the solution lies, we need to add these properties to this global override field. So I would expect us to add a few `<ModuleName>.fileExtensions` to the list with each module's specific version that it defines. Most likely all may be the same as Module's own version number.
> 
> Ideally, we could remove the need for this list if we did version comparing between the version the module was introduced versus the version the inherited property was introduced and taking the newest of the 2. We are tracking inherited properties directly through `SUPER_CLASS_PROPERTIES_JAVADOCS`. I would expect this to be done in another PR. @romani If you agree we should create the issue for this.
> 
> Edit: We would still need the override to fall back on if we changed extension of the module after it was introduced, like changing AbstractCheck to AbstractJavadocCheck. We have a few modules we are looking to do that to still. Only way to avoid the override is if we put a `since` version on the class hierarchy.

Sure @rnveach Sir , letme add <ModuleName>.fileextensions first. Then we will discuss furthur

**Comment by romani** [comment]
ci failure:
```
[INFO] Running com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.296 s <<< FAILURE! -- in com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
[ERROR] com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest -- Time elapsed: 2.296 s <<< ERROR!
java.lang.IllegalStateException: Exception was thrown while processing src/xdocs/checks/whitespace/typecastparenpad.xml.template
  at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
 ....
  at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
Caused by: java.lang.NullPointerException
  at com.puppycrawl.tools.checkstyle.utils.JavadocUtil.getFirstChild(JavadocUtil.java:220)
  at com.puppycrawl.tools.checkstyle.utils.JavadocUtil.findFirstToken(JavadocUtil.java:200)
  at com.puppycrawl.tools.checkstyle.site.SiteUtil.getSinceVersionFromJavadoc(SiteUtil.java:727)
  at com.puppycrawl.tools.checkstyle.site.SiteUtil.getSinceVersion(SiteUtil.java:707)
  at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertySinceVersionCell(PropertiesMacro.java:492)
  at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertyRow(PropertiesMacro.java:233)
  at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writeTablePropertiesRows(PropertiesMacro.java:187)
  at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertiesTable(PropertiesMacro.java:130)
  at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.execute(PropertiesMacro.java:102)
  at org.apache.maven.doxia.parser.AbstractParser.executeMacro(AbstractParser.java:145)
  at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.processMacroEnd(XdocsTemplateParser.java:187)
  at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.handleEndTag(XdocsTemplateParser.java:116)
  at org.apache.maven.doxia.parser.AbstractXmlParser.parseXml(AbstractXmlParser.java:238)
  at org.apache.maven.doxia.parser.AbstractXmlParser.parse(AbstractXmlParser.java:156)
  at org.apache.maven.doxia.parser.XhtmlBaseParser.parse(XhtmlBaseParser.java:92)
  at org.apache.maven.doxia.module.xdoc.XdocParser.parse(XdocParser.java:113)
  at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.parse(XdocsTemplateParser.java:76)
  at org.apache.maven.doxia.parser.AbstractParser.parse(AbstractParser.java:207)
  at com.puppycrawl.tools.checkstyle.internal.utils.XdocGenerator.generateXdocContent(XdocGenerator.java:68)
  ... 62 more

```

**Comment by Rohanraj123** [comment]
> ci failure:
> 
> ```
> [INFO] Running com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
> [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 2.296 s <<< FAILURE! -- in com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest
> [ERROR] com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest -- Time elapsed: 2.296 s <<< ERROR!
> java.lang.IllegalStateException: Exception was thrown while processing src/xdocs/checks/whitespace/typecastparenpad.xml.template
>   at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
>  ....
>   at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
> Caused by: java.lang.NullPointerException
>   at com.puppycrawl.tools.checkstyle.utils.JavadocUtil.getFirstChild(JavadocUtil.java:220)
>   at com.puppycrawl.tools.checkstyle.utils.JavadocUtil.findFirstToken(JavadocUtil.java:200)
>   at com.puppycrawl.tools.checkstyle.site.SiteUtil.getSinceVersionFromJavadoc(SiteUtil.java:727)
>   at com.puppycrawl.tools.checkstyle.site.SiteUtil.getSinceVersion(SiteUtil.java:707)
>   at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertySinceVersionCell(PropertiesMacro.java:492)
>   at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertyRow(PropertiesMacro.java:233)
>   at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writeTablePropertiesRows(PropertiesMacro.java:187)
>   at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.writePropertiesTable(PropertiesMacro.java:130)
>   at com.puppycrawl.tools.checkstyle.site.PropertiesMacro.execute(PropertiesMacro.java:102)
>   at org.apache.maven.doxia.parser.AbstractParser.executeMacro(AbstractParser.java:145)
>   at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.processMacroEnd(XdocsTemplateParser.java:187)
>   at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.handleEndTag(XdocsTemplateParser.java:116)
>   at org.apache.maven.doxia.parser.AbstractXmlParser.parseXml(AbstractXmlParser.java:238)
>   at org.apache.maven.doxia.parser.AbstractXmlParser.parse(AbstractXmlParser.java:156)
>   at org.apache.maven.doxia.parser.XhtmlBaseParser.parse(XhtmlBaseParser.java:92)
>   at org.apache.maven.doxia.module.xdoc.XdocParser.parse(XdocParser.java:113)
>   at com.puppycrawl.tools.checkstyle.site.XdocsTemplateParser.parse(XdocsTemplateParser.java:76)
>   at org.apache.maven.doxia.parser.AbstractParser.parse(AbstractParser.java:207)
>   at com.puppycrawl.tools.checkstyle.internal.utils.XdocGenerator.generateXdocContent(XdocGenerator.java:68)
>   ... 62 more
> ```

I fixed this CI failure @romani Sir.

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java [hunk 2]
```diff
@@ -699,9 +702,7 @@ public static String getSinceVersion(String moduleName, DetailNode moduleJavadoc
         if (superClassSinceVersion != null) {
             sinceVersion = superClassSinceVersion;
         }
-        else if (SUPER_CLASS_PROPERTIES_JAVADOCS.containsKey(propertyName)
-                || TOKENS.equals(propertyName)
-                || JAVADOC_TOKENS.equals(propertyName)) {
+        else if (TOKENS.equals(propertyName) || JAVADOC_TOKENS.equals(propertyName)) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java:705]:
````suggestion
        else if (TOKENS.equals(propertyName)
                || JAVADOC_TOKENS.equals(propertyName)) {
````

**Review by rnveach — CHANGES_REQUESTED** [review]
Should be final change

File: src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java (hunk 2 above)
Comment (Rohanraj123) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java:705]: Sure sir

Comment (Rohanraj123) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: @rnveach Even After making these changes it is not updating sinceVersion.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: @Rohanraj123 , restore formatting only
```
else if (TOKENS.equals(propertyName)
                || JAVADOC_TOKENS.equals(propertyName)) {
```

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: code looks good.

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: > Even After making these changes it is not updating sinceVersion.

What do you mean?
It was updated at https://github.com/checkstyle/checkstyle/pull/13945/files#diff-eaff5fa3cb101145757a2bd9ebe2e805e8fd8057ae27f9e21ef91900cb7938cc

Comment (Rohanraj123) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: Okay got it thanks !

Comment (Rohanraj123) [reply src/main/java/com/puppycrawl/tools/checkstyle/site/SiteUtil.java]: > @Rohanraj123 , restore formatting only
> 
> ```
> else if (TOKENS.equals(propertyName)
>                 || JAVADOC_TOKENS.equals(propertyName)) {
> ```

did it ! @romani Sir

**Review by rnveach — APPROVED** [review]

**Pull request merged**